libcaf_core/caf/async/spsc_buffer.hpp

```diff
@@ -164,24 +157,24 @@ class spsc_buffer : public ref_counted {
     if (consumer_)
       CAF_RAISE_ERROR("SPSC buffer already has a consumer");
     consumer_ = std::move(consumer);
-    flags_.had_consumer = true;
     if (producer_)
       ready();
-    else if (flags_.had_producer)
+    else if (flags_.closed)
       consumer_->on_producer_wakeup();
   }
 
   /// Producer callback for the initial handshake between producer and consumer.
   void set_producer(producer_ptr producer) {
     CAF_ASSERT(producer != nullptr);
     lock_type guard{mtx_};
-    if (producer_)
-      CAF_RAISE_ERROR("SPSC buffer already has a producer");
+    if (producer_) {
+      producer->on_consumer_cancel();
+      return;
```

Comment:
Why does this propagate the error while calling `set_consumer` twice raises and error?

Reply:
Yeah, good question. I went a bit back-and-forth on that one. We simply have no way to "gracefully" abort a consumer. But having the producer behave differently is also odd. So for consistency: ack, reverting to raise an error in both cases. Also throwing a `std::logic_error` here instead of `std::runtime_error`.